File: examples/11_smoothing.py
"""11 - Smoothing: uniform vs cotangent Laplacian.

Both pull a perturbed vertex back toward its neighbourhood; cotangent weights
respect the surface metric so the result is independent of triangulation.
Boundary vertices stay fixed. Run:
    PYTHONPATH=build/bin python examples/11_smoothing.py
"""
import simplemesh as sm


def grid(n=6):
    m = sm.Mesh()
    v = [[m.add_vertex(sm.Vec3(x, y, 0)) for x in range(n)] for y in range(n)]
    for y in range(n - 1):
        for x in range(n - 1):
            m.add_triangle(v[y][x], v[y][x + 1], v[y + 1][x + 1])
            m.add_triangle(v[y][x], v[y + 1][x + 1], v[y + 1][x])
    return m, v


def spike(m, v):
    p = m.point(v); m.set_point(v, sm.Vec3(p.x, p.y, 1.0))   # lift one vertex

m, v = grid(); mid = v[3][3]
spike(m, mid)
sm.laplacian_smooth(m, iterations=10, lambda_=0.5)
print("uniform  Laplacian: spike z =", round(m.point(mid).z, 4))

m, v = grid(); mid = v[3][3]
spike(m, mid)
sm.cotan_smooth(m, iterations=10, lambda_=0.5)
print("cotangent Laplacian: spike z =", round(m.point(mid).z, 4))

File: examples/11_smoothing.cpp
// 11 - Smoothing: uniform vs cotangent Laplacian (C++).
//   Both pull a perturbed vertex back toward its neighbourhood; boundary fixed.
#include "simplemesh/Mesh.h"
#include "simplemesh/Algorithms.h"
#include <iostream>
#include <vector>

static sm::Mesh grid(int n, sm::VertexHandle& mid) {
    sm::Mesh m;
    std::vector<std::vector<sm::VertexHandle>> v(n, std::vector<sm::VertexHandle>(n));
    for (int y = 0; y < n; ++y)
        for (int x = 0; x < n; ++x)
            v[y][x] = m.add_vertex({double(x), double(y), 0});
    for (int y = 0; y < n - 1; ++y)
        for (int x = 0; x < n - 1; ++x) {
            m.add_triangle(v[y][x], v[y][x+1], v[y+1][x+1]);
            m.add_triangle(v[y][x], v[y+1][x+1], v[y+1][x]);
        }
    mid = v[n/2][n/2];
    return m;
}

int main() {
    sm::VertexHandle mid;
    {
        sm::Mesh m = grid(6, mid);
        sm::Vec3 p = m.point(mid); p.z = 1.0; m.set_point(mid, p);  // spike
        sm::laplacian_smooth(m, 10, 0.5);
        std::cout << "uniform   Laplacian: spike z = " << m.point(mid).z << "\n";
    }
    {
        sm::Mesh m = grid(6, mid);
        sm::Vec3 p = m.point(mid); p.z = 1.0; m.set_point(mid, p);
        sm::cotan_smooth(m, 10, 0.5);
        std::cout << "cotangent Laplacian: spike z = " << m.point(mid).z << "\n";
    }
    return 0;
}
